### PR TITLE
WIP: POC of Timetable feature. Includes AbstractTimetable and  CronTimetable classes.

### PR DESCRIPTION
**Pull request by jhtimmins**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/models/dag.py [hunk 1]
```diff
@@ -109,6 +110,312 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+        start_date=None,
+        end_date=None,
+    ):
+        self.schedule_interval = schedule_interval
+        self.timezone = timezone
+        self.catchup = catchup
+        self.start_date = start_date
+        self.end_date = end_date
+
+    def normalized_schedule_interval(self) -> Optional[ScheduleInterval]:
```
Comment (dimberman) [inline airflow/models/dag.py:242]: this function name feels a bit vague to me. I would ad docs to the abstract class or consider making it mroe descriptive.

File: airflow/models/dag.py [hunk 2]
```diff
@@ -109,6 +110,312 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+        """
+        Returns Normalized Schedule Interval. This is used internally by the Scheduler to
+        schedule DAGs.
+
+        1. Converts Cron Preset to a Cron Expression (e.g ``@monthly`` to ``0 0 1 * *``)
+        2. If Schedule Interval is "@once" return "None"
+        3. If not (1) or (2) returns schedule_interval
+        """
+        if isinstance(self.schedule_interval, str) and self.schedule_interval in cron_presets:
```
Comment (dimberman) [inline airflow/models/dag.py:251]: Will this be a necessary step for anyone looking to backport? Is there a way we can make this easier?

File: airflow/models/dag.py [hunk 3]
```diff
@@ -109,6 +110,312 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+
+        return False
+
+    def date_range(
+        self,
+        start_date: datetime,
+        num: Optional[int] = None,
+        end_date: Optional[datetime] = timezone.utcnow(),
+    ) -> List[datetime]:
+        if num is not None:
```
Comment (dimberman) [inline airflow/models/dag.py:282]: num is not a descriptive variable name.

File: airflow/models/dag.py [hunk 4]
```diff
@@ -109,6 +110,312 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+        :return: utc datetime
+        """
+        if isinstance(self.normalized_schedule_interval(), str):
+            # we don't want to rely on the transitions created by
+            # croniter as they are not always correct
+            dttm = pendulum.instance(dttm)
+            naive = timezone.make_naive(dttm, self.timezone)
+            cron = croniter(self.normalized_schedule_interval(), naive)
+
+            # We assume that DST transitions happen on the minute/hour
```
Comment (dimberman) [inline airflow/models/dag.py:302]: Is that a reasonable assumption to make?

File: airflow/models/dag.py [hunk 5]
```diff
@@ -109,6 +110,312 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+        """
+        Calculates the previous schedule for this dag in UTC
+
+        :param dttm: utc datetime
+        :return: utc datetime
+        """
+        if isinstance(self.normalized_schedule_interval(), str):
+            # we don't want to rely on the transitions created by
+            # croniter as they are not always correct
+            dttm = pendulum.instance(dttm)
```
Comment (dimberman) [inline airflow/models/dag.py:326]: these names are not descriptive

**Review by dimberman — CHANGES_REQUESTED** [review]

File: airflow/models/dag.py (hunk 1 above)
Comment (jhtimmins) [reply airflow/models/dag.py:242]: @dimberman  This code is almost entirely copy+pasted from existing code. So I'll make things readable before sending it around.

File: airflow/models/dag.py (hunk 2 above)
Comment (ashb) [reply airflow/models/dag.py:251]: Backport what?

File: airflow/models/dag.py (hunk 4 above)
Comment (ashb) [reply airflow/models/dag.py:302]: Not a new assumption -- this is just moving the existing code around.

So far, yes, this is a reasonable assumption though, yes

File: airflow/models/dag.py [hunk 6]
```diff
@@ -109,6 +110,308 @@ def get_last_dagrun(dag_id, session, include_externally_triggered=False):
...
+        if next_run_date and self.end_date and next_run_date > self.end_date:
+            return None
+
+        # Don't schedule a dag beyond its end_date (as specified by the task params)
+        # Get the min task end date, which may come from the dag.default_args
+        task_end_dates = [t.end_date for t in tasks if t.end_date]
+        if task_end_dates and next_run_date:
+            min_task_end_date = min(task_end_dates)
+            if next_run_date > min_task_end_date:
+                return None
```
Comment (ashb) [inline airflow/models/dag.py:442]: Some of this feels like it should be in the base timetable class -- or at least, each timetable subclass shouldn't be responsible for implementing this check etc.

**Pull request closed without merging**